### PR TITLE
Add Server GraphQL to `site`

### PR DESCRIPTION
**Pull request by heowc**
Motivation:

There is no documentation for Server GraphQL

Modifications:

- Add `server-graphql`

Result:

- Add Server GraphQL to `site`
- Relates #3863

### CODE REVIEW

**Comment by heowc** [comment]
스크린샷 2021-10-20 오후 8 57 54

File: site/src/pages/docs/server-graphql.mdx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        final String id = environment.getArgument("id");
+        return data.get(id);
+    }
+}
+```
+<Tip>
+
+Please note that this is entirely based on [graphql-java](https://www.graphql-java.com/).
+
+</Tip>
```
Comment (ikhoon) [inline site/src/pages/docs/server-graphql.mdx:49]: Could we add simple explanations for schema and DataFetcher?

---

[GraphQL Java]((https://www.graphql-java.com/) is the Java (server) implementation for GraphQL.
This documentation shows how to serve a GraphQL Java over HTTP.
The main steps for creating a GraphQL service are:
1. Defining a GraphQL schema.
2. Implement how the actual data for a query is fetched.
3. Bind the scheme and data fetchers with `GraphqlService`.

## Defining Schema

A schema which defines each field that can be queried or mutated and what types those fields are.
This schema defines one top level field (in the type Query): which returns the details of a User.
```graphql
type Query {
  user(id: ID): User
}

type User {
  id: ID
  name: String
}
```

## Loading Data
A `DataFetcher` fetches data for one field while the query is executed.
The `UserDataFetcher` retrieves a user with a specific ID from a static map. This is made just for testing.
```
class UserDataFetcher implements DataFetcher<User> {
    ...
}
```

## Building `GraphqlService`

...

Comment (heowc) [reply site/src/pages/docs/server-graphql.mdx]: I've added it mixed up with what I've written. 👍👍

Comment (heowc) [reply site/src/pages/docs/server-graphql.mdx]: localhost_8000_docs_server-graphql

File: site/src/pages/docs/server-graphql.mdx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+- [GraphQL Java](https://www.graphql-java.com/) is the Java (server) implementation for GraphQL.
+- This documentation shows how to serve a service made from GraphQL Java over HTTP.
+- The main steps for creating a GraphQL service are:
+
+## Defining schema
+
+A schema which defines each field that can be queried or mutated and what types those fields are.
+This schema defines one top level field (in the type Query): which returns the details of a User.
+
+```graphql
```
Comment (minwoox) [inline site/src/pages/docs/server-graphql.mdx:33]: How about?

Let's assume we have the following [GraphQL Schema](https://graphql.org/learn/schema/):

```graphql
type Query {
  user(id: ID): User
}

type User {
  id: ID
  name: String
}
```

The schema describes that if a client requests a `User` data using the `ID` field, the GraphQL server will
return the `User` data. The GraphQL engine uses a [DataFetcher](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/DataFetcher.html)
to resolve the `ID` and fetch the corresponding object. For the sake of simplicity, let's implement
`UserDataFetcher` that stores `User` data in a `Map`:

```java
import graphql.schema.DataFetcher;
import graphql.schema.DataFetchingEnvironment;

class UserDataFetcher implements DataFetcher<User> {

    private final Map<String, User> data = Map.of("1", new User("1", "hero"),
                                                  "2", new User("2", "human"),
                                                  "3", new User("3", "droid"));

    @Override
    public User get(DataFetchingEnvironment environment) throws Exception {
        final String id = environment.getArgument("id");
        return data.get(id);
    }
}
```

## `GraphqlService`
Once you've finished the implementation of the `DataFetcher`, you need to build a <type://GraphqlService> using
a <type://GraphqlServiceBuilder> and add it to the <type://ServerBuilder>:

```java
import com.linecorp.armeria.server.Server;
import com.linecorp.armeria.server.ServerBuilder;

ServerBuilder sb = Server.builder();
...
sb.service("/graphql",
           GraphqlService.builder()
                         .runtimeWiring(c -> {
                            c.type("Query",
                                   typeWiring -> typeWiring.dataFetcher("user", new UserDataFetcher()));
                         })
                         .build());
...
Server server = sb.build();
server.start();
```

// explanation about runtimeWiring

## Blocking service implementation

...

Comment (minwoox) [reply site/src/pages/docs/server-graphql.mdx:33]: @freevie you might want to take a look at this PR. 😄

Comment (heowc) [reply site/src/pages/docs/server-graphql.mdx:11]: I'm sorry for being late. 😅 I fixed

File: site/src/pages/docs/server-graphql.mdx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+Server server = sb.build();
+server.start();
+```
+
+Here [RuntimeWiring](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/idl/RuntimeWiring.html) is a specification of [DataFetcher](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/DataFetcher.html)s, [TypeResolver](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/TypeResolver.html)s, and custom [scalar](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/GraphQLScalarType.html)s that are needed to wire together a functional [GraphQLSchema](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/GraphQLSchema.html).
+
+## Blocking service implementation
+
+Armeria does not run service logic in a separate thread pool by default. If your service implementation
+requires blocking, either run the individual blocking logic in a thread pool, set
```
Comment (minwoox) [inline site/src/pages/docs/server-graphql.mdx:79]:
```suggestion
requires blocking, either run the individual blocking logic in a thread pool, or set
```

File: site/src/pages/docs/server-graphql.mdx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+<type://GraphqlServiceBuilder#useBlockingTaskExecutor(boolean)> so the above happens automatically for
```
Comment (minwoox) [inline site/src/pages/docs/server-graphql.mdx:80]:
```suggestion
<type://GraphqlServiceBuilder#useBlockingTaskExecutor(boolean)> to `true` so the service runs in
the <type://CommonPools#blockingTaskExecutor()>.
```

File: site/src/pages/docs/server-graphql.mdx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                         })
+                         // All service methods will be run within
+                         // the blocking executor.
+                         .useBlockingTaskExecutor(true)
+                         .build());
+```
+
+<Tip>
+
+Another way is to wrap the [DataFetcher](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/DataFetcher.html) in an [AsyncDataFetcher](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/AsyncDataFetcher.html).
```
Comment (minwoox) [inline site/src/pages/docs/server-graphql.mdx:99]: How about:
You can wrap a DataFetcher using AsyncDataFetcher] to run it asynchronously.

File: site/src/pages/docs/server-graphql.mdx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                            c.type("Query",
+                                   typeWiring -> typeWiring.dataFetcher("user", new UserDataFetcher()));
+                         })
+                         .build());
+...
+Server server = sb.build();
+server.start();
+```
+
+Here [RuntimeWiring](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/idl/RuntimeWiring.html) is a specification of [DataFetcher](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/DataFetcher.html)s, [TypeResolver](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/TypeResolver.html)s, and custom [scalar](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/GraphQLScalarType.html)s that are needed to wire together a functional [GraphQLSchema](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/GraphQLSchema.html).
```
Comment (minwoox) [inline site/src/pages/docs/server-graphql.mdx:74]: How about:
We used [RuntimeWiring](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/idl/RuntimeWiring.html) to wire the type with DataFetcher. Please see [...](https://www.graphql-java.com/documentation/master/schema/#creating-a-schema-using-the-sdl) to find more wiring examples.

**Review by minwoox — COMMENTED** [review]
Thanks! Left a few more nits. 😉

**Review by minwoox — APPROVED** [review]
Thanks a lot, @heowc! 😄

File: site/src/pages/docs/server-graphql.mdx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                                                  "3", new User("3", "droid"));
+
+    @Override
+    public User get(DataFetchingEnvironment environment) throws Exception {
+        final String id = environment.getArgument("id");
+        return data.get(id);
+    }
+}
+```
+
```
Comment (ikhoon) [inline site/src/pages/docs/server-graphql.mdx:44]: Could you add the dependency section for `armeria-graphql`? For example:
https://github.com/line/armeria/blob/14078714659814a8d34cc9fbca15ee011524c4dc/site/src/pages/docs/advanced-zipkin.mdx#L9-L16

File: site/src/pages/docs/server-graphql.mdx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                            c.type("Query",
+                                   typeWiring -> typeWiring.dataFetcher("user", new UserDataFetcher()));
+                         })
+                         .build());
+...
+Server server = sb.build();
+server.start();
+```
+
+We used [RuntimeWiring](https://javadoc.io/doc/com.graphql-java/graphql-java/latest/graphql/schema/idl/RuntimeWiring.html) to wire the type with DataFetcher. Please see [...](https://www.graphql-java.com/documentation/master/schema/#creating-a-schema-using-the-sdl) to find more wiring examples.
```
Comment (ikhoon) [inline site/src/pages/docs/server-graphql.mdx:75]: Is `[...]` cruft or intentionally added? A proper link text would be good to me.

Comment (minwoox) [reply site/src/pages/docs/server-graphql.mdx:75]: Oops, it should be `[Creating a schema using the SDL]` 😅

File: site/src/pages/docs/server-graphql.mdx (hunk 7 above)
Comment (heowc) [reply site/src/pages/docs/server-graphql.mdx:53]: Added again 🤣

File: site/src/pages/docs/server-graphql.mdx (hunk 8 above)
Comment (heowc) [reply site/src/pages/docs/server-graphql.mdx]: Fixed 🤣

**Review by ikhoon — APPROVED** [review]
Thanks a lot, @heowc!

**Review by jrhee17 — APPROVED** [review]
💯 Looks great! Thank you @heowc 🎉 🙇 🥳

**Comment by minwoox** [comment]
Thanks, @heowc for improving our documentation. 😄 🎉

**Pull request merged**